### ISSUE TITLE
AMP error for @custom-media

### ISSUE THREAD

**Issue by teddarling** [wprig/wprig#400]
## Issue Overview
The use of @custom-media is not allowed with AMP.

## Describe your environment
Using VS Code, Chrome/Safari/Firefox.

## Steps to Reproduce
1. Build WP Rig
2. Install WP Rig
3. Install AMP Plugin (https://wordpress.org/plugins/amp/)Install AMP Plugin (https://wordpress.org/plugins/amp/)
4. Activate AMP Plugin
5. Validate any page
6. View AMP Validation Errors
7. Notice result.

## Expected Behavior
A valid amp page on install of AMP and Built WP Rig setup.

## Current Behavior
AMP provides the following error:
- illegal_css_at_rule: @custom-media

## Possible Solution
Use either hard coded CSS media query values or use SASS to build the CSS with SASS variables.

## Screenshots / Video
![AMP Error](https://user-images.githubusercontent.com/3044359/55365120-82f36300-54a0-11e9-9518-877c984f0cfc.png)

**Comment by GitaStreet** [comment]
Yes, this issue also makes variables disappear from:
 _custom-properties.css

Commented this line out in various .css files - and the variable definitions are back:
@import "_custom-media.css

**Comment by GitaStreet** [comment]
Found this suggestion that seems to work:
"CSS variables should to be set in :root { }"
vuejs-templates/webpack#413

Quickfix not thoroughly tested:

Uncomment or delete all instances of:
@import "_custom-media.css";
Move these lines of code into the :root{} in the file _custom-properties.css
/* stylelint-disable */ @custom-media --narrow-menu-query screen and (max-width: 37.5em); @custom-media --wide-menu-query screen and (min-width: 37.5em); @custom-media --content-query screen and (min-width: 48em); @custom-media --sidebar-query screen and (min-width: 60em); /* stylelint-enable */

Error message still there, but theme appears to render correctly.

**Comment by teddarling** [comment]
First, Let me add I updated my initial post to indicate that this is in version 2.0

@GitaStreet, I did add the variables into custom-properties.css, as you mentioned.

```
	/* stylelint-disable */
	@custom-media --narrow-menu-query screen and (max-width: 37.5em);
	@custom-media --wide-menu-query screen and (min-width: 37.5em);
	@custom-media --content-query screen and (min-width: 48em);
	@custom-media --sidebar-query screen and (min-width: 60em);
	/* stylelint-enable */
```

While this removed the AMP validation error, it doesn't actually work with media queries. For example, the `.menu-toggle` under `@media (--narrow-menu-query)` inside global.css, doesn't handle showing the menu when the class on the menu button is changed. However, just changing `@media (--narrow-menu-query)` to `@media screen and (max-width: 37.5em)` does fire the media query, indicating that the variable isn't correct. I also tried the following with the same results.



```
	/* stylelint-disable */
	--narrow-menu-query: screen and (max-width: 37.5em);
	--wide-menu-query: screen and (min-width: 37.5em);
	--content-query: screen and (min-width: 48em);
	--sidebar-query: screen and (min-width: 60em);
	/* stylelint-enable */
```

So again, my statement stands that if you want to use variables for media queries, that it's probably best to use SASS to get the CSS to be AMP compatible (and functional). Or, code directly into media queries where needed (implementing the suggestion in #303 would make this easier to manage).

**Comment by westonruter** [comment]
Yeah, `@custom-media` is not standard CSS, so it is illegal in AMP. Why is it getting served to the browser in the first place? Browsers won't understand it either. The SASS should always be compiled before being loaded into a page.

**Comment by GitaStreet** [comment]
@teddarling 
Trying to understand. Seems to be working here.

@westonruter 
`@custom-media` appears to be compiled. 

Have an WP RIG and AMP running here:

[https://zugest.dk/](https://zugest.dk/)

**Comment by benoitchantre** [comment]
In `config.json`, set the value of `importFrom`to `["_custom-properties.css", "_custom-media.css"]`.
This will force postCSS to compute the values.

**Comment by mor10** [comment]
Custom media is bleeding edge. I'm starting to think we should shelve it for now in favor of standard variables.

**Comment by teddarling** [comment]
ok, so I started from fresh, pulled version 2 into a clean install.

Here are my settings:

config.json:
```
{
  "dev": {
    "styles": {
      "importFrom": ["_custom-properties.css", "_custom-media.css"]
    }
  }
}
```

The I run `npm run rig-init` followed by `npm run build`

Here is the first line of global.min.css file that is generated

```
@custom-media --narrow-menu-query screen and (max-width: 37.5em);@custom-media --wide-menu-query screen and (min-width: 37.5em);@custom-media --content-query screen and (min-width: 48em);@custom-media --sidebar-query screen and (min-width: 60em);:root{--global-font-color:#333;--global-font-family:"Crimson Text",serif;--global-font-size:20;--global-font-line-height:1.4;--highlight-font-family:"Roboto Condensed",sans-serif;--content-width:45rem;--dropdown-symbol-width:0.7em;--color-theme-primary:#e36d60;--color-theme-secondary:#41848f;--color-theme-red:#c0392b;--color-theme-green:#27ae60;--color-theme-blue:#2980b9;--color-theme-yellow:#f1c40f;--color-theme-black:#1c2833;--color-theme-grey:#95a5a6;--color-theme-white:#ecf0f1;--color-custom-daylight:#97c0b7;--color-custom-sun:#eee9d1;--color-link:#0073aa;--color-link-visited:#333;--color-link-active:#00a0d2;--color-quote-border:#000;--color-quote-citation:#6c7781;--font-size-small:calc(16/var(--global-font-size)*1rem);--font-size-regular:calc(var(--global-font-size)/16*1rem);--font-size-large:calc(36/var(--global-font-size)*1rem);--font-size-larger:calc(48/var(--global-font-size)*1rem)}
```

Notice, the very first line has @custom-media, thus the AMP validation errors.

No I will say, that the build is building out with the custom media queries, but I would say that this is something in and of itself that is going to hurt the CSS limit within AMP generated CSS.

So first ( with config set to debug css true), this is an example of the output global.min.css file.

```
@media screen and (max-width: 37.5em) {

	.nav--toggle-small.nav--toggled-on .menu {
		display: block;
	}
}

@media (--narrow-menu-query) {

	.nav--toggle-small.nav--toggled-on .menu {
		display: block;
	}
}
```

So this is duplicated CSS, and AMP couldn't possibly handle tree shaking on this item, so now we have an issue where all @custom-media and custom properties are generating duplicated content.

I've also noticed, that while the generated CSS does include :root{}:

```
:root {
	--global-font-color: #333;
	--global-font-family: "Crimson Text", serif;
	--global-font-size: 20;
	--global-font-line-height: 1.4;
	--highlight-font-family: "Roboto Condensed", sans-serif;
	--content-width: 45rem;
	--dropdown-symbol-width: 0.7em;

	/* Custom editor colors */
	--color-theme-primary: #e36d60;
	--color-theme-secondary: #41848f;
	--color-theme-red: #c0392b;
	--color-theme-green: #27ae60;
	--color-theme-blue: #2980b9;
	--color-theme-yellow: #f1c40f;
	--color-theme-black: #1c2833;
	--color-theme-grey: #95a5a6;
	--color-theme-white: #ecf0f1;
	--color-custom-daylight: #97c0b7;
	--color-custom-sun: #eee9d1;

	--color-link: #0073aa;
	--color-link-visited: #333;
	--color-link-active: #00a0d2;

	--color-quote-border: #000000;
	--color-quote-citation: #6c7781;

	/* Custom editor font sizes */
	--font-size-small: calc(16 / var(--global-font-size) * 1rem);
	--font-size-regular: calc(var(--global-font-size) / 16 * 1rem);
	--font-size-large: calc(36 / var(--global-font-size) * 1rem);
	--font-size-larger: calc(48 / var(--global-font-size) * 1rem);
}
```

the AMP changes to the CSS removes it from `<style amp-custom>` section, thus the variables are not even present for something like `font-family: var(--global-font-family);` This then causes font display issues, because what AMP process as the final result is this:

```
body, button, input {
    color: #333;
    color: var(--global-font-color);
    font-family: "Crimson Text",serif;
    font-family: var(--global-font-family);
    font-size: calc(20 / 16 * 1rem);
    font-size: var(--font-size-regular);
    line-height: 1.4;
    line-height: var(--global-font-line-height);
}
```

Where `font-family: var(--global-font-family);` is what the browser tries to process, overriding the first font-family declaration. Thus, for example, Crimson Text is not displayed and instead it defaults to the browser base font.

![Screen Shot 2019-04-02 at 10 04 50 AM](https://user-images.githubusercontent.com/3044359/55417911-d7452400-552e-11e9-983c-4f77a0a16bd6.png)

**Comment by westonruter** [comment]
AMP currently allows the following at-rules:

* `@font-face`
* `@keyframes`
* `@media`
* `@supports`
* `@page`

The AMP plugin automatically drops `@charset` and `@import` in the process of concatenating into the `style[amp-custom]` stylesheet. Now, just because AMP doesn't support an at-rule doesn't mean that it it was intentionally disallowed. For example, I added `@page` support in https://github.com/ampproject/amphtml/pull/17481. So if a case can be made for other such at-rules which don't violate [AMP's design principles](https://www.ampproject.org/about/amp-design-principles/) (e.g. negative implications for security or performance).

**Comment by teddarling** [comment]
@mor10 I do unserstand the bleeding edge. It's the main reason that I'm looking at using WP Rig. That along with the built in AMP integration.

I think the biggest issue is that the bleeding edge that is available for browsers is different than for AMP.

But perhaps my image included above can show you what AMP compiles to with what is generated in the build process. That image alone (bottom of https://github.com/wprig/wprig/issues/400#issuecomment-479069988) shows that font-family, font-size and line-height are all overridden by variables that don't exist in the amp-custom style.

So what I've seen, is that what looks good, if I invalidate or disable a page from being processed by AMP, looks and behaves completely different once the AMP plugin has done its thing to the CSS.

Perhaps the solution is to have two different builds with different CSS enqueued for AMP than if no AMP is enabled for a page, or as you said, maybe shelve it for now. But do be aware that it appears CSS variables are not available in what AMP generates as well.

@westonruter, I know there is a lot of movement toward CSS variables. It appears that the AMP plugin removes those as well. Is that part of what isn't included in AMP?

**Comment by westonruter** [comment]
@teddarling The AMP plugin does not remove CSS variables. Using a test plugin like https://gist.github.com/westonruter/11fe27a17f2d38dc331407cdba699a67 which outputs:

```html
<style>
    :root {
        --content-width: 640px;
    }
    body {
        max-width: var(--content-width);
    }
</style>
```

The `style[amp-custom]` generated by the plugin includes:

```css
:root{--content-width:640px}body{max-width:var(--content-width)}
```

So the variables are intact. CSS variables are valid in AMP, and the AMP plugin does not remove them.

The issue is actually with the [PHP-CSS-Parser](https://github.com/sabberworm/PHP-CSS-Parser) the plugin relies on. What appears is happening is that when it comes across CSS like this:

```css
@custom-media --narrow-menu-query screen and (max-width: 37.5em);
:root {
    --content-width: 640px;
}
body {
    max-width: var(--content-width);
}
```

It is failing to parse the `@custom-media` CSS rule as:

```css
@custom-media --narrow-menu-query screen and (max-width: 37.5em);
:root {
    --content-width: 640px;
}
```

So then the resulting CSS that the AMP plugin outputs is:

```css
body {
    max-width: var(--content-width);
}
```

So this is why the `:root` rule is getting unexpectedly dropped. The issue is `@custom-media`. It's not allowed in AMP, and the PHP-CSS-Parser doesn't understand it to boot.

So the fix for AMP compatibility is to eliminate `@custom-media`.

**Comment by mor10** [comment]
@teddarling See #403: I am proposing removing the bundled `@custom-media` rules for now while leaving the functionality in place. Eventually `@custom-media` will have better support, in AMP and outside. Until then my suggestion is we'll mark it as experimental and allow people to make their own decisions.

**Comment by benoitchantre** [comment]
@teddarling Can you test #405?

**Comment by ataylorme** [comment]
Closed by #405
